tests/test_TFNetworkLayer.py

```diff
@@ -150,6 +150,141 @@ def test_concat_sources_missing_dim():
+
+      "att_value_window": {"class": "window", "window_size": context_size, "padding": "valid", "from": "att_v_split"},
+      "att_out_dot": {"class": "dot", "red1": "f", "red2": "s:1", "var1": None, "var2": "f", "from": ["att_scales", "att_value_window"]},
+
+      "att_out_combine": {"class": "merge_dims", "axes": ["s:1", "f"], "from": "att_out_dot"},
+      "output": {"class": "copy", "from": ["att_out_combine"]}
+    }
+
+    config = Config({"debug_print_layer_output_template": True})
+    in0 = Data(
+      name="in0", shape=(None, n_in // channels, channels), batch_dim_axis=0, auto_create_placeholders=True)
+    extern_data = ExternData()
+    extern_data.register_data(in0)
+
+    network = TFNetwork(config=config, extern_data=extern_data, train_flag=True)
+    network.construct_from_dict(net_dict)
+
+    in_data = rnd.normal(size=(n_batch, n_time, n_in // channels, channels)).astype("float32")
+    in_seq_len = numpy.empty(n_batch)
+    in_seq_len.fill(n_time)
+
+    feed_dict = {in0.placeholder: in_data,
+                 in0.size_placeholder[0]: in_seq_len
+    }
+    fetches = []
+
+    for _, l in network.layers.items():
+      l.output.placeholder = tf.Print(
+          l.output.placeholder, [l.name, "shape:", tf.shape(l.output.placeholder)],
+          summarize=10, name="debug_print_layer_output_shape")
+      fetches.append(l.output.placeholder)
+    output = network.get_default_output_layer().output
+
+    tf.global_variables_initializer().run()
+    out = session.run(fetches=fetches,
+      feed_dict=feed_dict)
+    inp1, inp2 = network.layers["att_energy_dot"].sources[0].output.placeholder, network.layers["att_energy_dot"].sources[1].output.placeholder
+    res = network.layers["att_energy_dot"].output.placeholder
+    inp1, inp2, res = session.run([inp1, inp2, res], feed_dict=feed_dict)
+
+
+def test_CombineLayer():
+  import numpy as np
+  from TFNetwork import TFNetwork, ExternData
+  with make_scope() as session:
+    net = TFNetwork(extern_data=ExternData())
+    src_data = Data(name='t', shape=(None, 32, 128), dtype='float32', auto_create_placeholders=True)
+    window_size = 5
+    slice_size = window_size - 1
+
+    rnd = np.random.RandomState(42)
+    input_data = rnd.rand(10, 11, 32, 128)
+    seq_lens = np.array([11, 11, 11, 11, 11, 11, 11, 11, 11, 11])
+
+    with tf.variable_scope("src"):
+      src = InternalLayer(name="src", network=net, out_type={"dim": 128,
+                                                             "shape": (None, 32, 128),
+                                                             "batch_dim_axis": 0,
+                                                             "time_dim_axis": 1,
+                                                             "feature_dim_axis": 3,
+                                                             "sparse": False
+                                                             })
```

Comment:
Please don't format the code in this way. Never indent so far to the right. Rather, just use new line, such that every further indent is *always* exactly 2 spaces.
Also, for debugging purpose, it would be helpful to have sth like this:
```print("src:", src)```
(The same also for the following layers.)

Reply:
Done